### ISSUE TITLE
Using MDRangePolicy in Hierarchical Parallelism 

### ISSUE THREAD

**Issue by kbthomp1** [kokkos/kokkos#3385]
It would be nice to be able to use MDRangePolicy within a TeamPolicy to exploit hierarchical parallel with loop tiling on each team's threads.  The specific use case is for structured grid problems where there are multiple grid blocks of different sizes, preventing a tightly nested 4D loop.  Being able to use an MDRangePolicy on each team allows each block to be assigned a team that then uses a 3D MDRangePolicy for the tightly nested loops on each structured grid block.

**Comment by dalg24** [comment]
Tentative 3.5 release

**Comment by crtrott** [comment]
Something like:

```
parallel_for(TeamPolicy<>(...), KOKKOS_LAMBDA(const team_t& team) {
  ....
  parallel_for(TeamVectorMDRange<Iterate::Left>(team, N0, N1, N2) , [=] (int i, int j, int k) {
     ....
  });
  ...
});
```
Essentially the equivalents to our TeamThreadRange, ThreadVectorRange, and TeamVectorRange just with MD. No tiling needed here.

**Comment by kbthomp1** [comment]
@crtrott Yes exactly.  Something that provides MD versions of TeamThreadRange, ThreadVectorRange, and TeamVectorRange, but no explicit tiling needed.  Preferably, with the default template `Iterate` parameter switching from `Iterate::Right` to `Iterate::Left` depending on the default execution space, similar to how MDRangePolicy does.

**Comment by crtrott** [comment]
Yeah we were just prioritizing work and put this on the list for the next release. I just added a bit more description to make clear what the plan is.

**Comment by vbrunini** [comment]
Did this make it into 3.5? Guessing not?

**Comment by masterleinad** [comment]
> Did this make it into 3.5? Guessing not?

No.

**Comment by PhilMiller** [comment]
Delivered in the just-released 4.0.00
